clients/filesystem-fuse/src/filesystem.rs

```diff
@@ -0,0 +1,236 @@
+    /// Init the file system
+    async fn init(&self);
+
+    /// Get the file stat by file path
+    async fn stat(&self, name: &str) -> Result<FileStat>;
+
+    /// Get the file stat by parent file path and file name
+    async fn lookup(&self, parent: &str, name: &str) -> Result<FileStat>;
+
+    /// Read the directory by file path
+    async fn read_dir(&self, name: &str) -> Result<Vec<FileStat>>;
+
+    /// Open the file by file path and flags
+    async fn open_file(&self, name: &str, flags: OpenFileFlags) -> Result<OpenedFile>;
+
+    /// Open the directory by file path and flags
+    async fn open_dir(&self, name: &str, flags: OpenFileFlags) -> Result<OpenedFile>;
+
+    /// Create the file by parent file path and file name and flags
+    async fn create_file(
+        &self,
+        parent: &str,
+        name: &str,
+        flags: OpenFileFlags,
+    ) -> Result<OpenedFile>;
+
+    /// Create the directory by parent file path and file name
+    async fn create_dir(&self, parent: &str, name: &str) -> Result<OpenedFile>;
+
+    /// Set the file attribute by file path and file stat
+    async fn set_attr(&self, name: &str, file_stat: &FileStat, flush: bool) -> Result<()>;
+
+    /// Remove the file by parent file path and file name
+    async fn remove_file(&self, parent: &str, name: &str) -> Result<()>;
+
+    /// Remove the directory by parent file path and file name
+    async fn remove_dir(&self, parent: &str, name: &str) -> Result<()>;
+}
+
+pub(crate) struct FileSystemContext {
+    // system user id
+    pub(crate) uid: u32,
+
+    // system group id
+    pub(crate) gid: u32,
+
+    // default file permission
+    pub(crate) default_file_perm: u16,
+
+    // default idr permission
+    pub(crate) default_dir_perm: u16,
+
+    // io block size
+    pub(crate) block_size: u32,
+}
+
+// FileSystemContext is the file system context for the file system implementation.
+impl FileSystemContext {
+    pub(crate) fn new(uid: u32, gid: u32) -> Self {
+        FileSystemContext {
+            uid,
+            gid,
+            default_file_perm: 0o644,
+            default_dir_perm: 0o755,
+            block_size: 4 * 1024,
+        }
+    }
+}
+
+// FileStat is the file stat for the file system.
+#[derive(Clone, Debug)]
+pub struct FileStat {
+    // file id for the file system.
+    pub(crate) file_id: u64,
+
+    // parent file id
+    pub(crate) parent_file_id: u64,
+
+    // file name
+    pub(crate) name: String,
+
+    // file path of the fuse file system root
+    pub(crate) path: String,
+
+    // file size
+    pub(crate) size: u64,
+
+    // file type like regular file or directory and so on
+    pub(crate) kind: FileType,
+
+    // file permission
+    pub(crate) perm: u16,
+
+    // file access time
+    pub(crate) atime: Timestamp,
+
+    // file modify time
+    pub(crate) mtime: Timestamp,
+
+    // file create time
+    pub(crate) ctime: Timestamp,
+
+    // file link count
+    pub(crate) nlink: u32,
```

Comment:
do we need `nlink`?

Reply:
That's a reference count of the file. we may use it later.

Reply:
Could you provide an example about how to use it ?

Reply:
This can be considered in a later PR. Let's first push the framework so others can join in.